### PR TITLE
[dashboard] Redesign of dashboard

### PR DESCRIPTION
**Pull request by svenefftinge**
We are reimplementing and redesigning the Dashboard.

The current one is a bit grown and has some problems wrt maintenance and loading speed. Also we never really leveraged what material-ui provides us but carried the weight with us.

The new implementation shall provide a clean architecture, based on a react-router SPA. We are (DRI @gtsiolis) building out our own minimal design system based on TailwindCSS. 

For the first iteration the goal is to simply replace all the existing functionality. We explicitly don't add new features or change remodel existing ones heavily, so that the server API would need to change. We are however redesigning the UI.

### Technology & Architectural Patterns

The dashboard is written in TypeScript and React. For styling it uses TailwindCSS which is a bit nicer than inlining CSS as it supports pseudo classes and a is a little more abstract/reusable.

The App.tsx is the entry point for the SPA and it uses React-Router to register all pages. 

```ts
    <Switch>
        <Route path="/" exact component={Workspaces} />
        <Route path="/profile" exact component={Profile} />
        <Route path="/notifications" exact component={Notifications} />
        <Route path="/subscriptions" exact component={Subscriptions} />
        <Route path="/env-vars" exact component={EnvVars} />
        <Route path="/git-integration" exact component={GitIntegration} />
        <Route path="/feature-preview" exact component={FeaturePreview} />
        <Route path="/default-ide" exact component={DefaultIDE} />
    </Switch>
```
Pages are loaded lazily using `React.lazy` so that not everything needs to be loaded up-front but only when needed:

```ts
const Notifications = React.lazy(() => import('./account/Notifications'));
const Profile = React.lazy(() => import('./account/Profile'));
const Subscriptions = React.lazy(() => import('./account/Subscriptions'));
const DefaultIDE = React.lazy(() => import('./settings/DefaultIDE'));
const EnvVars = React.lazy(() => import('./settings/EnvVars'));
const FeaturePreview = React.lazy(() => import('./settings/FeaturePreview'));
const GitIntegration = React.lazy(() => import('./settings/GitIntegration'));
```

Global state is passed through `React.Context`.

Fixes https://github.com/gitpod-io/gitpod/issues/3300

### CODE REVIEW

File: components/dashboard/src/contexts.tsx [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+const UserContext = createContext<{
+    user?: User,
+    setUser: React.Dispatch<User>
+}>({
+    setUser: () => null
+});
+
+
+const AppProvider: React.FC = ({ children }) => {
```
Comment (svenefftinge) [inline components/dashboard/src/contexts.tsx:12]: UserContextProvider

File: components/dashboard/src/contexts.tsx [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { User } from '@gitpod/gitpod-protocol';
```
Comment (svenefftinge) [inline components/dashboard/src/contexts.tsx:1]: file name should be `user-context.tsx`

File: components/dashboard/src/service/service.ts [hunk 3]
```diff
@@ -52,4 +42,38 @@ function createGitpodService<C extends GitpodClient, S extends GitpodServer>() {
...
+            this.userPromise = this.gitpodService.server.getLoggedInUser();
+        }
+        return this.userPromise;
+    }
+    async reloadUser() {
+        this.userPromise = undefined;
+        return this.getOrLoadUser();
+    }
+
+    async getAuthProviders() {
```
Comment (svenefftinge) [inline components/dashboard/src/service/service.ts:63]: Is there a reason for this indirection?

File: components/dashboard/src/service/service.ts [hunk 4]
```diff
@@ -52,4 +42,38 @@ function createGitpodService<C extends GitpodClient, S extends GitpodServer>() {
...
+        return this.gitpodService.server.getAuthProviders();
+    }
+
+}
+
+let gitpodService: GitpodService;
+let service: AppService;
```
Comment (svenefftinge) [inline components/dashboard/src/service/service.ts:70]: how about storing the userPromise here instead, we might be able to go without an `AppService`.

File: components/dashboard/src/index.tsx [hunk 5]
```diff
@@ -1,18 +1,18 @@
...
-const service = new SimpleServiceImpl();
-service.service.server.getLoggedInUser().then(user => {
-  service.user = user;
-  ReactDOM.render(
+ReactDOM.render(
     <React.StrictMode>
-      <ServiceContext.Provider value={service}>
-        <App />
-      </ServiceContext.Provider>
+        <AppProvider>
```
Comment (svenefftinge) [inline components/dashboard/src/index.tsx:13]: We could pass in the `userPromise` as a prop. Seems like this is the only client for that. So no need to store it globally.

File: components/dashboard/src/service/service.ts [hunk 6]
```diff
@@ -4,25 +4,14 @@
...
-    workspaces: WorkspaceInfo[];
-    service: GitpodService;
-}
-
-export class SimpleServiceImpl implements Service {
-    user?: User;
-    service = createGitpodService();
-    workspaces: WorkspaceInfo[] = [];
-}
+export const gitpodHostUrl = new GitpodHostUrl((window as any).PREVIEW_URL || window.location.toString());
```
Comment (svenefftinge) [inline components/dashboard/src/service/service.ts:14]: I don't think we should build such a variance into the code. It should always be `window.location` or manually changed here (and not checked in).

File: components/dashboard/public/index.html [hunk 7]
```diff
@@ -24,6 +24,9 @@
       work correctly both with client-side routing and a non-root public URL.
       Learn how to configure a non-root public URL by running `npm run build`.
     -->
+    <script>
+      window.PREVIEW_URL = "%REACT_APP_PREVIEW_URL%";
+    </script>
```
Comment (svenefftinge) [inline components/dashboard/public/index.html:32]: Let's not make this a variance. Just hard coding it temporarily seems fine and simple.

**Comment by jankeromnes** [comment]
Werft failure sounds like a race:

```
error Couldn't publish package: "https://registry.yarnpkg.com/@gitpod%2fsupervisor-api-grpc: Failed to save packument, a common cause is if you try to publish a new package before the previous package has been fully processed."
```

Re-trying:

/werft run

**Comment by geropl** [comment]
/werft run

:crossed_fingers:

**Review by gtsiolis — APPROVED** [review]
Alright, LET'S DO THIS.

**Review by gtsiolis — APPROVED** [review]
Alright, LET'S DO THIS.

**Pull request merged**